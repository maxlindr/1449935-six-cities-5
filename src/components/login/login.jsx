import React, {memo} from 'react';
import PropTypes from 'prop-types';
import PageHeader from '../page-header/page-header';
import withLoginFormController from '../../hocs/with-login-form-controller/with-login-form-controller';
import './animation.css';
import withRedirectAuthorized from '../../hocs/with-redirect-authorized/with-redirect-authorized';

const ANIMATION_CLASSNAME = `shake`;

const Login = (props) => {
  const {
    email,
    password,
    isValid,
    isDisabled,
    isAnimationPlaying,
    onEmailChange,
    onPasswordChange,
    onSubmit,
  } = props;

  const formClassName = isAnimationPlaying
    ? `login__form form ${ANIMATION_CLASSNAME}`
    : `login__form form`;

  return (
    <div className="page page--gray page--login">
      <PageHeader />

      <main className="page__main page__main--login">
        <div className="page__login-container container">
          <section className="login">
            <h1 className="login__title">Sign in</h1>

            <form
              className={formClassName}
              action="#"
              method="post"
              onSubmit={onSubmit}
            >
              <div className="login__input-wrapper form__input-wrapper">
                <label className="visually-hidden">E-mail</label>
                <input
                  className="login__input form__input"
                  type="email"
                  name="email"
                  placeholder="Email"
                  required
                  disabled={isDisabled}
                  value={email}
                  onChange={onEmailChange}
                />
              </div>

              <div className="login__input-wrapper form__input-wrapper">
                <label className="visually-hidden">Password</label>
                <input
                  className="login__input form__input"
                  type="password"
                  name="password"
                  placeholder="Password"
                  required
                  disabled={isDisabled}
                  value={password}
                  onChange={onPasswordChange}
                />
              </div>

              <button
                className="login__submit form__submit button"
                type="submit"
                disabled={!isValid || isDisabled}
              >
                Sign in
              </button>
            </form>
          </section>

          <section className="locations locations--login locations--current">
            <div className="locations__item">
              <a className="locations__item-link" href="#">
                <span>Amsterdam</span>
              </a>
            </div>
          </section>
        </div>
      </main>
    </div>
  );
};

Login.propTypes = {
  email: PropTypes.string.isRequired,
  password: PropTypes.string.isRequired,
  isValid: PropTypes.bool.isRequired,
  isDisabled: PropTypes.bool.isRequired,
  isAnimationPlaying: PropTypes.bool.isRequired,
  onEmailChange: PropTypes.func.isRequired,
  onPasswordChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
};

export {Login};

export default withRedirectAuthorized(
    withLoginFormController(
        memo(Login)
    )
);
