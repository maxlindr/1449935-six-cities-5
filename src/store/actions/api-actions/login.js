import {APIRoute, AppRoute, AuthorizationStatus, ErrorMessage, HttpCode} from '../../../constants';
import {ActionCreator} from '../action';

export const login = (email, password) => (dispatch, _getState, api) => {
  dispatch(ActionCreator.setAuthorizationStatus(AuthorizationStatus.PENDING));

  api.post(APIRoute.LOGIN, {email, password})
    .then(({data}) => {
      dispatch(ActionCreator.setAuthorizationStatus(AuthorizationStatus.AUTHORIZED));
      dispatch(ActionCreator.setUserData(data));
      dispatch(ActionCreator.redirectToRoute(AppRoute.ROOT));
    })
    .catch((err) => {
      dispatch(ActionCreator.setAuthorizationStatus(AuthorizationStatus.NOT_AUTHORIZED));

      if (
        err.response &&
        err.response.status === HttpCode.NOT_AUTHORIZED
      ) {
        dispatch(ActionCreator.setLoginFailed(true));
      } else {
        dispatch(ActionCreator.setErrorMessage(ErrorMessage.GENERAL));
        dispatch(ActionCreator.goToRoute(AppRoute.ERROR));
      }
    });
};
