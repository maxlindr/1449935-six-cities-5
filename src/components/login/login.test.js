import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import {Login} from './login';

const renderer = new ShallowRenderer();

test(`Login должен корректно отображаться`, () => {
  renderer.render(
      <Login
        email={`Oliver.conner@gmail.com`}
        password={`123456`}
        isValid={true}
        isDisabled={false}
        isAnimationPlaying={false}
        onEmailChange={() => {}}
        onPasswordChange={() => {}}
        onSubmit={() => {}}
      />
  );

  const result = renderer.getRenderOutput();
  expect(result).toMatchSnapshot();
});
