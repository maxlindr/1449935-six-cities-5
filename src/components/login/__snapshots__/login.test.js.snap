// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login должен корректно отображаться 1`] = `
<div
  className="page page--gray page--login"
>
  <Memo(Connect(PageHeader)) />
  <main
    className="page__main page__main--login"
  >
    <div
      className="page__login-container container"
    >
      <section
        className="login"
      >
        <h1
          className="login__title"
        >
          Sign in
        </h1>
        <form
          action="#"
          className="login__form form"
          method="post"
          onSubmit={[Function]}
        >
          <div
            className="login__input-wrapper form__input-wrapper"
          >
            <label
              className="visually-hidden"
            >
              E-mail
            </label>
            <input
              className="login__input form__input"
              disabled={false}
              name="email"
              onChange={[Function]}
              placeholder="Email"
              required={true}
              type="email"
              value="Oliver.conner@gmail.com"
            />
          </div>
          <div
            className="login__input-wrapper form__input-wrapper"
          >
            <label
              className="visually-hidden"
            >
              Password
            </label>
            <input
              className="login__input form__input"
              disabled={false}
              name="password"
              onChange={[Function]}
              placeholder="Password"
              required={true}
              type="password"
              value="123456"
            />
          </div>
          <button
            className="login__submit form__submit button"
            disabled={false}
            type="submit"
          >
            Sign in
          </button>
        </form>
      </section>
      <section
        className="locations locations--login locations--current"
      >
        <div
          className="locations__item"
        >
          <a
            className="locations__item-link"
            href="#"
          >
            <span>
              Amsterdam
            </span>
          </a>
        </div>
      </section>
    </div>
  </main>
</div>
`;
